core/src/graph/graph.rs

```diff
@@ -245,51 +245,107 @@ impl Graph {
-					None => match self.get_next_available_page_id() {
-						Some(next_page_id) => Some(Box::new(GraphPage::new(
-							self.get_connection_type().privacy_type(),
-							next_page_id,
-						))),
-						None => None,
-					},
+		// Now go through the remaining connections to be added and see if we can
+		// add them to other existing pages that are non-full. Here we prefer to only
+		// aggressively scan pages for fullness, because we want to minimize the number
+		// of additional pages to be updated.
+		let updated_keys: HashSet<PageId> = updated_pages.keys().cloned().collect();
+		'existing_page_loop: for (_, page) in
+			self.pages.inner().iter().filter(|(page_id, _)| !updated_keys.contains(page_id))
+		{
+			let mut current_page = page.clone();
+			let mut page_modified = false;
+			while let Some(id_to_add) = add_iter.peek() {
+				if let Ok(_) = self.try_add_connection_to_page(
+					&mut current_page,
+					id_to_add,
+					true,
+					&encryption_key,
+					Some(dsnp_version_config),
+				) {
+					page_modified = true;
+					let _ = add_iter.next(); // TODO: prefer advance_by(1) once stabilized
+						 // If no more connections to add, we're done. Otherwise we'll continue
+						 // adding connections to the current page until either it's full, or
+						 // we have no more connections to add.
+					if let None = add_iter.peek() {
+						updated_pages.insert(current_page.page_id(), current_page);
+						break 'existing_page_loop
+					}
+				}
+				// If we couldn't add a connection to the current page, it's full. Add this page
+				// to the updated pages list and move on to the next page.
+				else {
+					if page_modified {
+						updated_pages.insert(current_page.page_id(), current_page.clone());
+					}
+					continue 'existing_page_loop
+				}
+			}
+		}
+
+		// At this point, all existing pages are aggressively full. Add new pages
+		// as needed to accommodate any remaining connections to be added, filling aggressively.
+		let mut new_page: Option<Box<GraphPage>> = None;
+		while let Some(id_to_add) = add_iter.peek() {
+			if new_page.is_none() {
+				if let Some(next_page_id) = self.get_next_available_page_id() {
+					new_page = Some(Box::new(GraphPage::new(
+						self.get_connection_type().privacy_type(),
+						next_page_id,
+					)));
+				} else {
+					return Err(Error::msg("Graph is full; no new pages available"))
 				}
 			}
 
-			match current_page {
-				Some(ref mut page) => {
-					page.add_connection(add_iter.next().unwrap())?; // safe to unwrap because we peeked above
-					if self.is_page_full(&page, false) {
+			if let Some(page) = new_page.as_mut() {
+				if self
+					.try_add_connection_to_page(
+						&mut **page,
+						id_to_add,
+						true,
+						&encryption_key,
+						Some(dsnp_version_config),
+					)
+					.is_ok()
+				{
+					let _ = add_iter.next(); // TODO: prefer advance_by(1) once stabilized
+					if add_iter.peek().is_none() {
```

Comment:
do you really need this if you are doing the if Some check on line 345?

Reply:
Good point.